### PR TITLE
Willem/genesis ledger implementation

### LINKED ISSUE #170: M4 - ledger serialization/deserialization
**Task summary**

* Store & Update
Introduce the ability to store and update account database. Convert ledger data structure into a format that is transferable via network, graphql api or disk storage.

* <del>Transition frontier</del>. Unstable as of now.
<del>Maintain transition frontier and other required state objects</del>

* Bindings to query
Provide JavaScript bindings to query account state

**Specification reference**

- [Code](https://github.com/MinaProtocol/mina/blob/fedbf62c4f37d89ca921152f4dacdd65c9b31e18/src/lib/genesis_ledger_helper/genesis_ledger_helper.ml) to create or load genesis ledger from disk, 
- Genesis ledger [data structure](https://github.com/MinaProtocol/mina/blob/fedbf62c4f37d89ca921152f4dacdd65c9b31e18/src/lib/genesis_ledger/genesis_ledger.ml) and [tests](https://github.com/MinaProtocol/mina/blob/fedbf62c4f37d89ca921152f4dacdd65c9b31e18/src/lib/test_genesis_ledger/test_genesis_ledger.ml)
- [Code](https://github.com/ChainSafe/mina-rs/pull/162/files#diff-fd0293ea1784f925a76c6cd0f336b9d20d05a451b6c228dc221d4a15b14c070dR182) to add additional instrumentation

**Other information and links**

### PR DESCRIPTION
**Pull request by willemolding**
# Genesis Ledger

!! Review and merge #182 first!!

The genesis ledger allows a new Mina instance to set-up the accounts that should exist from genesis. Mina-OCaml stores the genesis ledgers for mainnet and testnet as RocksDB instances and so the simplest way if we want to be able to get the same initial ledger is to be able to read and ingest these genesis ledgers.

**Summary of changes**
Changes introduced in this pull request:
- Adds `GenesisLedger` trait
- Adds an implementation of `GenesisLedger` that uses RocksDB as a back-end and allows for reading the Mina genesis ledgers

**Reference issue to close (if applicable)**

Closes #170 
Depends #182 

**Other information and links**

### CODE REVIEW

File: base/src/serialization_type_conversions/account.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                timing: t.timing.into(),
+                permissions: t.permissions.into(),
+                snapp: t.snapp.map(Into::into),
+            },
+        )))
+    }
+}
+impl From<AccountV1> for Account {
+    fn from(t: AccountV1) -> Self {
+        Self {
```
Comment (hanabi1224) [inline base/src/serialization_type_conversions/account.rs:32]: Nit: Maybe `let t = t.t.t.t;` to save some bits

**Review by hanabi1224 — APPROVED** [review]

File: ledger/src/genesis_ledger.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    /// Error type that can be produces when trying to access the underlying store
+    type Error;
+
+    /// Return the depth of the ledger
+    fn depth(&self) -> usize {
+        DEPTH
+    }
+
+    /// Return a iterator over the accounts in this genesis ledger without consuming self
+    fn accounts(&'a self) -> <&'a Self as IntoIterator>::IntoIter {
```
Comment (hanabi1224) [inline ledger/src/genesis_ledger.rs:31]: another `fn hashes(&'a self) -> IntoIterator<Item = (usize, LedgerHash)` fn will be super helpful

File: ledger/src/rocksdb_genesis_ledger.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        Box::new(db_iter.map(decode_account_from_kv))
+    }
+}
+
+impl<'a, const DEPTH: usize> GenesisLedger<'a, DEPTH> for RocksDbGenesisLedger<'a, DEPTH> {
+    type Error = Error;
+}
+
+#[cfg(test)]
```
Comment (creativcoder) [inline ledger/src/rocksdb_genesis_ledger.rs:66]: `#[cfg(not(target_arch = "wasm32"))]`

File: ledger/src/rocksdb_genesis_ledger.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+/// A genesis ledger backed by a RocksDB instance
+pub struct RocksDbGenesisLedger<'a, const DEPTH: usize> {
+    db: &'a DB,
+}
+
+/// Errors than can be produces when trying to access a Rocksdb backed ledger
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Could not deserialize account")]
+    Disconnect(#[from] bin_prot::error::Error),
```
Comment (creativcoder) [inline ledger/src/rocksdb_genesis_ledger.rs:26]: The variant type and the error description does not concur

**Review by creativcoder — APPROVED** [review]
Left a few comments, thanks for the kickoff on ledger. Awesome start!

**Pull request merged**